crates/services/integration/src/audiobookshelf.rs

```diff
@@ -0,0 +1,163 @@
+use media_models::{IntegrationMediaCollection, IntegrationMediaSeen};
+use providers::google_books::GoogleBooksService;
+use specific_models::audiobookshelf;
+
+use crate::integration::Integration;
+
+pub struct AudiobookshelfIntegration {
+    base_url: String,
+    access_token: String,
+    isbn_service: GoogleBooksService,
+}
+
+impl AudiobookshelfIntegration {
+    pub fn new(base_url: String, access_token: String, isbn_service: GoogleBooksService) -> Self {
+        Self {
+            base_url,
+            access_token,
+            isbn_service,
+        }
+    }
+}
+
+impl Integration for AudiobookshelfIntegration {
+    async fn progress(
+        &self,
+    ) -> anyhow::Result<(Vec<IntegrationMediaSeen>, Vec<IntegrationMediaCollection>)> {
+        let client = get_base_http_client(
+            &format!("{}/api/", self.base_url),
+            Some(vec![(
+                AUTHORIZATION,
+                HeaderValue::from_str(&format!("Bearer {}", self.access_token)).unwrap(),
+            )]),
+        );
+
+        let resp = client
+            .get("me/items-in-progress")
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?
+            .json::<audiobookshelf::Response>()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        tracing::debug!("Got response for items in progress {:?}", resp);
+
+        let mut media_items = vec![];
+
+        for item in resp.library_items.iter() {
+            let metadata = item.media.clone().unwrap().metadata;
+            let (progress_id, identifier, lot, source, podcast_episode_number) =
+                if Some("epub".to_string()) == item.media.as_ref().unwrap().ebook_format {
+                    match &metadata.isbn {
+                        Some(isbn) => match self.isbn_service.id_from_isbn(isbn).await {
+                            Some(id) => (
+                                item.id.clone(),
+                                id,
+                                MediaLot::Book,
+                                MediaSource::GoogleBooks,
+                                None,
+                            ),
+                            _ => {
+                                tracing::debug!("No Google Books ID found for ISBN {:#?}", isbn);
+                                continue;
+                            }
+                        },
+                        _ => {
+                            tracing::debug!("No ISBN found for item {:#?}", item);
+                            continue;
+                        }
+                    }
+                } else if let Some(asin) = metadata.asin.clone() {
+                    (
+                        item.id.clone(),
+                        asin,
+                        MediaLot::AudioBook,
+                        MediaSource::Audible,
+                        None,
+                    )
+                }
+                // else if let Some(itunes_id) = metadata.itunes_id.clone() {
```

Comment:
Why was this commented?

Reply:
Its still being worked on. Ive just finished getting it working and havent pushed it yet passing the closure required some changes.

Reply:
Itll take a little longer we have conflicting changes now ive got to resolve them first

Reply:
I accidently introduced 15 merge conflicts and my client handled it really poorly so im gonna need to make these changes again from scratch ill finish it tomorrow.